File: bindings/lua/examples/callback.lua
local source = debug.getinfo(1, "S").source:sub(2)
local dir = source:match("^(.*)/[^/]+$") or "."
package.path = dir .. "/../?.lua;" .. package.path

local urb = require("urb_ffi")
local ffi, mem = urb.ffi, urb.memory

local libc = ffi.open("libc.so.6")
local qsort_desc = ffi.describe("void qsort(pointer, u64, u64, pointer)")
local cmp_desc = ffi.describe("i32 cmp(pointer, pointer)")
local qsort = ffi.bind(ffi.sym(libc, "qsort"), qsort_desc)

local nums = { 42, 7, 99, -3, 15 }
local n = #nums
local buf = mem.alloc(n * 4)
for i = 1, n do
    mem.writei32(buf + (i - 1) * 4, nums[i])
end

io.write("antes:")
for i = 1, n do io.write(i == 1 and " " or ", ", nums[i]) end
io.write("\n")

local cmp = ffi.callback(cmp_desc, function(a, b)
    local va = mem.readi32(a)
    local vb = mem.readi32(b)
    if va < vb then return -1 end
    if va > vb then return 1 end
    return 0
end)

qsort(buf, n, 4, cmp.ptr)

io.write("depois:")
for i = 1, n do
    io.write(i == 1 and " " or ", ", mem.readi32(buf + (i - 1) * 4))
end
io.write("\n")

mem.free(buf)
ffi.close(libc)

File: bindings/node/examples/callback.js
'use strict';
const { ffi, memory: mem } = require('../');

// qsort via callback — JS comparator → C function pointer
const libc = ffi.open('libc.so.6');
const qsortDesc = ffi.describe('void qsort(pointer, u64, u64, pointer)');
const cmpDesc = ffi.describe('i32 cmp(pointer, pointer)');
const qsort = ffi.bind(ffi.sym(libc, 'qsort'), qsortDesc);

const nums = [42, 7, 99, -3, 15];
const n    = nums.length;
const buf  = mem.alloc(BigInt(n * 4));
for (let i = 0; i < n; i++) mem.writei32(buf + BigInt(i * 4), nums[i]);

console.log('antes:', nums);

const cmp = ffi.callback(cmpDesc, (a, b) => {
    const va = mem.readi32(a);
    const vb = mem.readi32(b);
    return va < vb ? -1 : va > vb ? 1 : 0;
});

qsort(buf, BigInt(n), 4n, cmp.ptr);

const sorted = [];
for (let i = 0; i < n; i++) sorted.push(mem.readi32(buf + BigInt(i * 4)));
console.log('depois:', sorted);

mem.free(buf);
ffi.close(libc);
